llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py

```diff
@@ -44,89 +55,161 @@
 )
 
 
-class CorrectiveRAGPack(BaseLlamaPack):
-    def __init__(self, documents: List[Document], tavily_ai_apikey: str) -> None:
-        """Init params."""
+class RetrieveEvent(Event):
+    """Retrieve event (gets retrieved nodes)."""
+
+    retrieved_nodes: List[NodeWithScore]
+
+
+class RelevanceEvalEvent(Event):
+    """Relevance evaluation event (gets results of relevance evaluation)."""
+
+    relevant_results: List[str]
+
+
+class TextExtractEvent(Event):
+    """Text extract event. Extracts relevant text and concatenates."""
+
+    relevant_text: str
+
+
+class QueryEvent(Event):
+    """Query event. Queries given relevant text and search text."""
+
+    relevant_text: str
+    search_text: str
+
+
+class CorrectiveRAGWorkflow(Workflow):
+    @step(pass_context=True)
+    async def ingest(self, ctx: Context, ev: StartEvent) -> Optional[StopEvent]:
+        """Ingest step (for ingesting docs and initializing index)."""
+        documents: Optional[List[Document]] = ev.get("documents")
+        tavily_ai_apikey: Optional[str] = ev.get("tavily_ai_apikey")
+
+        if any(i is None for i in [documents, tavily_ai_apikey]):
+            return None
+
         llm = OpenAI(model="gpt-4")
-        self.relevancy_pipeline = QueryPipeline(
+        ctx.data["relevancy_pipeline"] = QueryPipeline(
             chain=[DEFAULT_RELEVANCY_PROMPT_TEMPLATE, llm]
         )
-        self.transform_query_pipeline = QueryPipeline(
+        ctx.data["transform_query_pipeline"] = QueryPipeline(
             chain=[DEFAULT_TRANSFORM_QUERY_TEMPLATE, llm]
         )
 
-        self.llm = llm
-        self.index = VectorStoreIndex.from_documents(documents)
-        self.tavily_tool = TavilyToolSpec(api_key=tavily_ai_apikey)
+        ctx.data["llm"] = llm
+        ctx.data["index"] = VectorStoreIndex.from_documents(documents)
+        ctx.data["tavily_tool"] = TavilyToolSpec(api_key=tavily_ai_apikey)
 
-    def get_modules(self) -> Dict[str, Any]:
-        """Get modules."""
-        return {"llm": self.llm, "index": self.index}
+        return StopEvent()
 
-    def retrieve_nodes(self, query_str: str, **kwargs: Any) -> List[NodeWithScore]:
+    @step(pass_context=True)
+    async def retrieve(self, ctx: Context, ev: StartEvent) -> Optional[RetrieveEvent]:
```

Comment:
I guess this event assumes the index is already constructed. Should we raise an error if not? Or make the index / llm / tavilly tool part of the workflow constructor?

Reply:
let's add a check?